ext/background/background.js

```diff
@@ -171,16 +171,49 @@ async function attemptRecordInteraction(tab, session, force) {
     return Promise.resolve();
   }
   setChromeActivity(a, tab);
+
+  const merged = await mergeAttachment(session, a);
   return recordInteraction(
     {
       interaction_type: (_) => 'VIEWED',
       idempotency_key: `${moment().startOf('day')}::${tab.title}`,
-      attachment: a,
+      attachment: merged,
     },
     authorize(session)
   );
 }
 
+// Queries Range for existing attachments and merges the attachments based on a
+// provider's configured behavior.
+// Default behavior leaves the existing attachment fields alone, but will add
+// new ones.
+async function mergeAttachment(session, attachment) {
+  const dedupe = getProviderDedupe(attachment.provider);
+  if (dedupe == MERGE_BEHAVIOR.REPLACE) return attachment;
+
+  const activity = await listActivity(attachment.provider, authorize(session));
+  for (const a of activity.attachments) {
+    if (a.source_id != attachment.source_id) continue;
+
+    switch (dedupe) {
+      case MERGE_BEHAVIOR.MERGE_NEW:
+        return {
+          ...a,
+          ...attachment,
+        };
+      case MERGE_BEHAVIOR.MERGE_EXISTING:
+      default:
+        return {
+          ...attachment,
+          ...a,
+        };
```

Comment:
Since the server maintains existing values when a new value is not sent, you could make `MERGE_EXISTING` delete any fields found in `a` from `attachment` before sending, instead of sending copies of the existing values in `a` back to the server. It depends on what you want to happen in the event of different race conditions 🤔

Reply:
Done. I have an allow list to ensure that the attachment doesn't end up in a bad state:

`const ATTACHMENT_CORE = ['source_id', 'provider', 'org_id', 'type', 'origin'];`

The first 3 are required to associate the attachment with the existing one on the server and the last 2 are included to ensure that the attachment is updated correctly.